cytodynamics-test-container/src/test/java/com/linkedin/cytodynamics/test/TestDynamicLoad.java

```diff
@@ -245,6 +274,124 @@ public void testLoadAandB() {
+            .withIsolationLevel(IsolationLevel.NONE)
+            .addParentPreferredClassPattern("java.*")
+            .build())
+        .addParentRelationship(ParentRelationshipBuilder.builder()
+            .withParentClassLoader(parentClassLoaderB)
+            .withIsolationLevel(IsolationLevel.NONE)
+            .build())
+        .build();
+    Class<?> clazz = loader.loadClass(Object.class, TestInterfaceImpl.class.getName());
+    assertEquals(clazz.getClassLoader(), parentClassLoaderA);
+  }
+
+  @Test(description = "Given that there are multiple parent relationships, but only the last parent has a class, the "
+      + "class should be loaded by the last parent")
+  public void testMultipleParentsLoadFromLastParent() throws MalformedURLException {
+    URL testApiJarURL = getTestJarUri("api").toURL();
+    ClassLoader parentClassLoaderA = new URLClassLoader(new URL[]{testApiJarURL, getTestJarUri("a").toURL()}, null);
+    ClassLoader parentClassLoaderB = new URLClassLoader(new URL[]{testApiJarURL, getTestJarUri("b").toURL()}, null);
+    Loader loader = LoaderBuilder
+        .anIsolatingLoader()
+        .withOriginRestriction(OriginRestriction.allowByDefault())
+        // will load class from parent, so don't need any classpath
+        .withClasspath(Collections.emptyList())
+        .addParentRelationship(ParentRelationshipBuilder.builder()
+            .withParentClassLoader(parentClassLoaderA)
+            // using FULL so that the implementation class is not loaded from this parent
+            .withIsolationLevel(IsolationLevel.FULL)
+            .addParentPreferredClassPattern("java.*")
+            .build())
+        .addParentRelationship(ParentRelationshipBuilder.builder()
+            .withParentClassLoader(parentClassLoaderB)
+            // using NONE so that we can load the class from the parent
+            .withIsolationLevel(IsolationLevel.NONE)
+            .build())
+        .build();
+    Class<?> clazz = loader.loadClass(Object.class, TestInterfaceImpl.class.getName());
+    assertEquals(clazz.getClassLoader(), parentClassLoaderB);
+  }
+
+  /**
+   * Graph structure:
+   *
+   * (common parent  <------------ (partial delegation
+   *  classloader)                  classloader)
+   *          ^                     ^
+   *          |                    /
+   *          |                   /
+   *         FULL               NONE
+   *          |                 /
+   *          |                /
+   *         (child classloader)
+   *
+   * A more concrete use case for this is that the common parent would contain some shared API interfaces that need to
+   * be used by both the partial delegation classloader and the child classloader. The partial delegation classloader
+   * and the child classloader could be managed by different owners, so they would need to be separate and isolated.
+   *
+   * It should not be possible to build a cyclic graph with the cytodynamics builder APIs, so we are just considering an
```

Comment:
Is this assumed or enforced? Do we have a check like this somewhere:

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java#L124

Reply:
There is no explicit check, but I believe there is sufficient immutability to prevent it.
The parent can only be set when constructing the `ClassLoader`s. In order to introduce a loop, then `ClassLoader` `A` would need to be built with `ClassLoader` `B `as a parent, and vice versa. So `A` would only be buildable if `B` already exists, and `B` would only be buildable if `A` already exists, and we can't have both (unless someone uses reflection, but reflection can get around pretty much any validation anyways, including the Kafka validation above).
I will add a little more detail to the comment to clarify.